bot/slack/slack_bot.py

```diff
@@ -0,0 +1,161 @@
+        self.plugin_set = BotPluginSet(plugin_dir)
+
+    def run(self):
+        if self.slack_client.rtm_connect():
+            logger.info("Bot connected to slack")
+            self.post_welcome_message(random.choice(greetings))
+            self.async_func(self)
+            self.listen_for_messages()
+        else:
+            logger.error("Unable to connect to slack")
+            sys.exit(1)
+
+    def handle_command(self, message_text):
+        response = ""
+        message_tokens = message_text.split()
+        command = message_tokens[0]
+        parameters = message_tokens[1:len(message_tokens)]
+
+        if command == '!help':
+            response = "\nHelp is on it's way...try these:\n"
+            for plugin in self.plugin_set.enabled_plugins:
+                response += "\n{0} -- {1}".format(
+                    plugin['command_name'],
+                    plugin['help_text']
+                )
+        else:
+            # Loop through plugins and send message to it
+            plugin_command_found = False
+            for plugin in self.plugin_set.enabled_plugins:
+                if command == plugin['command_name']:
+                    plugin_command_found = True
+                    response += "\n" + plugin['plugin_class'].handle_command(parameters)
+                    continue
+
+            if plugin_command_found is False:
+                response = "Unknown command: " + command + ". Try !help"
+        return response
+
+    def parse_command(self, content):
+        # messages look like this:
+        # pwnbus: @mozdef !help
+        tokens = content.split('@' + self.bot_name)
+        command = tokens[1].strip()
+        return command
+
+    def handle_message(self, message):
+        channel = message['channel']
+        thread_ts = message['ts']
+        # If we're already in a thread, reply within that thread
+        if 'thread_ts' in message:
+            thread_ts = message['thread_ts']
+        content = message['content']
+        command = self.parse_command(content)
+
+        response = self.handle_command(command)
+        if response is not "":
+            self.post_thread_message(
+                text=response,
+                channel=channel,
+                thread_ts=thread_ts
+            )
+
+    def listen_for_messages(self):
+        while True:
+            for slack_message in self.slack_client.rtm_read():
+                message_type = slack_message.get('type')
+                if message_type == 'desktop_notification':
+                    self.handle_message(slack_message)
+            time.sleep(1)
+
+    def post_thread_message(self, text, channel, thread_ts):
+        self.slack_client.api_call(
+            "chat.postMessage",
+            as_user="true",
+            channel=channel,
+            text=text,
+            thread_ts=thread_ts
+        )
+
+    def post_attachment(self, message, channel, color):
+        if channel is None:
```

Comment:
Would it make sense to give the `channel` argument to `post_attachment` a default value of `None` so that callers don't have to specify it in what I assume will be the majority of calls?

Reply:
The post_attachment function is more of an internal function (which I suppose I can rename with an underscore to make that more obvious. The callers of this function (post_info_message, post_welcome_message etc at https://github.com/mozilla/MozDef/pull/1018/files#diff-46fed7650379c8a8432b6efca3b11a91R145) set a default value of None here, so we're doing this "up the chain".

Reply:
I renamed the post_attachment function to match the private function naming convention at https://github.com/mozilla/MozDef/pull/1018/commits/6d20570292dfebcfbb18f16ae9f98146f36d6be3